scanpy/experimental/pp/_highly_variable_genes.py

```diff
@@ -90,6 +92,47 @@ def clac_clipped_res_sparse(gene: int, cell: int, value: np.float64) -> np.float
     return residuals
 
 
+def _calculate_res_dense_vectorized(
+    matrix: np.ndarray[np.float64],
+    *,
+    sums_genes: np.ndarray[np.float64],
+    sums_cells: np.ndarray[np.float64],
+    sum_total: np.float64,
+    clip: np.float64,
+    theta: np.float64,
+    n_genes: int,  # TODO: delete if not used
+    n_cells: int,
+) -> np.ndarray[np.float64]:
+    # TODO: potentially a lot to rewrite here
+    # TODO: is there a better/more common way we use? e.g. with dispatching?
+    if isinstance(matrix, DaskArray):
+        mu = da.outer(sums_genes, sums_cells) / sum_total
+    else:
+        mu = np.outer(sums_genes, sums_cells) / sum_total
+
+    values = matrix.T
+
+    mu_sum = values - mu
+    pre_res = mu_sum / np.sqrt(mu + mu * mu / theta)
+
+    def custom_clip(x, clip_val):
+        x[x < -clip_val] = -clip_val
+        x[x > clip_val] = clip_val
+        return x
```

Comment:
Make a utility and use with the other implementation.  Also maybe a better name?  `dask_sparse_compat_clip`?